core/src/main/java/org/mule/runtime/core/internal/routing/Foreach.java

```diff
@@ -155,14 +158,24 @@ boolean validateExpression(CoreEvent event) {
     return result;
   }
 
-  TypedValue setCurrentValue(int batchSize, ForeachContext foreachContext) {
+  TypedValue setCurrentValue(int batchSize, ForeachContext foreachContext, CoreEvent event) {
     TypedValue currentValue;
     Iterator<TypedValue<?>> iterator = foreachContext.getIterator();
     if (batchSize > 1) {
       int counter = 0;
       List currentBatch = new ArrayList<>();
       while (iterator.hasNext() && counter < batchSize) {
-        currentBatch.add(iterator.next());
+        TypedValue managedValue;
+        TypedValue<?> typedValue = iterator.next();
+        if (typedValue.getValue() instanceof EventBuilderConfigurer) {
```

Comment:
this is quite similar to `ForeachRouter#createTypedValuePartToProcess`. Refactor.